sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -435,6 +439,14 @@ private[hive] class HiveClientImpl(
     getRawTableOption(dbName, tableName).map(convertHiveTableToCatalogTable)
   }
 
+  override def getCatalogAndHiveTableOption(
+      dbName: String,
+      tableName: String): Option[CatalogAndHiveTable] = withHiveState {
+    logDebug(s"Looking up $dbName.$tableName")
+    getRawTableOption(dbName, tableName)
+      .map(t => new CatalogAndHiveTableImpl(convertHiveTableToCatalogTable(t), t))
```

Comment:
I think this is still sub-optimal. Sometimes we just need the raw hive table and this has an extra `convertHiveTableToCatalogTable` invocation. How about we make it lazy?
```
class RawHiveTable(val rawTable: Object) {
  def toSparkTable: CatalogTable = ...
}
```

Reply:
Ok, I've added this in https://github.com/apache/spark/pull/36027/commits/c033dc22e3006688b8d1819acaddcfe49ffda125 and made the `toCatalogTable` conversion lazy.
As the `convertHiveTableToCatalogTable()` converter needs the Hive shim and uses the `HiveTable` class I still separated the `RawHiveTable` trait and `RawHiveTableImpl` class and the latter became an inner class of `HiveClientImpl` to get access to the converter.